### PR TITLE
Add pyspark dataframe_loader

### PR DESCRIPTION
**Pull request by DavidKatz-il**
Add `dataframe_loader` to support `inputs`.

### CODE REVIEW

**Comment by natekupp** [comment]
hey @sryza can you take a look at this one too?

**Comment by sryza** [comment]
Thanks for your contribution @DavidKatz-il !  I'm impressed with your thoroughness.

An important area to think through is that Dagster should be able to support any version of Spark.  I'm concerned that, by wrapping every individual option to every Spark input format in Dagster config, it will tie us too closely to specific Spark versions.  I.e. Spark might add or remove an option, and then it becomes unclear which Spark version Dagster's configuration should reflect.

Although it doesn't give us the same advantage of having all the options nicely display in the Dagit config editor, I think a safer approach would be to use permissive dictionaries for the set of options within each input format.  E.g.
```
'jdbc': {},
'orc': {},
```

This allows Spark itself to remain as the arbiter of what config options are available.

**Comment by DavidKatz-il** [comment]
Hey @sryza, thanks for reviewing.
Right now the user can add options that we didn't configure because we are using `Permissive`.
So you suggesting to put an empty `Permissive`?
e.g.
```   
'jdbc': Permissive(),
'orc': Permissive(),
...
```
If so, what about options that are required, e.g. 'path'.

**Comment by sryza** [comment]
Thought about it a little bit more, and I think you're right.  As long as we use Permissive, it's helpful to include the detailed config.

I just launched a build for the PR.

**Comment by flvndh** [comment]
Great contribution @DavidKatz-il! What do you think of adding support for [Apache Iceberg](https://iceberg.apache.org/) and [Delta Lake](https://delta.io/) as well ?

**Comment by DavidKatz-il** [comment]
Hi @flvndh, Does it require packages outside `pyspark`?

**Comment by flvndh** [comment]
@DavidKatz-il just jars packages : `org.apache.iceberg:iceberg-spark3-runtime:0.9.0` and `io.delta:delta-core_2.12:0.7.0`.

**Comment by DavidKatz-il** [comment]
@sryza what do you think?

**Comment by sryza** [comment]
This change looks ready to go in.  I think adding support for Iceberg and Delta Lake would be cool.  @DavidKatz-il - if you're interested in adding those to this PR, then great.  If you'd like me to merge this as-is, also great.

**Comment by DavidKatz-il** [comment]
I added the file_type 'other'.

e.g.
```
'other': {    
    'path': 'path_to_file',    
    'format': 'delta',    
    ...
}
```

File: python_modules/libraries/dagster-pyspark/dagster_pyspark_tests/test_types.py [hunk 1]
```diff
@@ -66,3 +74,28 @@ def return_df(_):
         assert result.success
         actual = read(temp_path)
         assert sorted(df.collect()) == sorted(actual.collect())
+
+
+@pytest.mark.parametrize(
```
Comment (sryza) [inline python_modules/libraries/dagster-pyspark/dagster_pyspark_tests/test_types.py:79]: would you mind adding tests for the "other" cases for input and output?

**Comment by sryza** [comment]
@schrockn brought up the question of whether we should avoid the "other" escape hatch in favor of making it easy to compose type materializers: https://dagster.phacility.com/D4188.

@DavidKatz-il - it could take some time for us to reach some resolution on that, but I still think the core of this PR, which adds support for the native Spark formats, is useful.  If you want to take out the "other" option for now, I can merge the rest of it.  Thanks for bearing with us on the thrash here.

File: python_modules/libraries/dagster-pyspark/dagster_pyspark_tests/test_types.py [hunk 2]
```diff
@@ -51,11 +69,9 @@ def return_df(_):
                         'outputs': [
                             {
                                 'df': {
-                                    file_type: {
-                                        'path': temp_path,
-                                        'mode': 'overwrite',
-                                        'compression': 'gzip',
-                                    }
+                                    file_type: dict(
+                                        {'mode': 'overwrite', 'compression': 'gzip',}, **options,
```
Comment (sryza) [inline python_modules/libraries/dagster-pyspark/dagster_pyspark_tests/test_types.py:73]: I'm noticing a syntax error here for in the python 2.7 tests: https://buildkite.com/dagster/dagster/builds/14227#9a8db1d9-518e-4491-87a8-d247ca90946c

Comment (DavidKatz-il) [reply python_modules/libraries/dagster-pyspark/dagster_pyspark_tests/test_types.py]: The comma at the end of the line caused this error.

**Comment by sryza** [comment]
This looks good.  Merging.  Thanks for your contribution @DavidKatz-il !

**Pull request merged**